File: Java/src/com/dbc/algorithm/Array/PascalsTriangle.java
package com.dbc.algorithm.Array;

import java.util.ArrayList;
import java.util.List;

public class PascalsTriangle {
    public static List<List<Integer>> generate(int numRows) {
        List<List<Integer>> result = new ArrayList<>();
        if (numRows == 0) {
            return result;
        }
        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<>();
            row.add(1);
            if (i == 0) {
                result.add(row);
                continue;
            }
            List<Integer> last = result.get(i - 1);
            for (int j = 1; j < i; j++) {
                row.add(last.get(j) + last.get(j - 1));
            }
            row.add(1);
            result.add(row);
        }
        return result;
    }
}

File: Python/array/pascals_triangle.py
from typing import List

def generate(numRows: int) -> List[List[int]]:
    if numRows == 0:
        return []
    result = [[1]]
    for i in range(numRows - 1):
        row = result[i].copy()
        for j in range(1, i + 1):
            if j == 1:
                temp = row[j]
                row[j] = row[j] + row[j - 1]
            else:
                remain = row[j]
                row[j] = row[j] + temp
                temp = remain
        
        row.append(1)
        result.append(row)
    
    return result

if __name__ == "__main__":
    print(generate(5))